### PR TITLE
Swift3.0

### PR DESCRIPTION
**Pull request by ffried**
### New Pull Request Checklist

* [x] I have read and understood the [CONTRIBUTING guide](https://github.com/CocoaLumberjack/CocoaLumberjack/blob/master/.github/CONTRIBUTING.md)
* [x] I have read the [Documentation](http://cocoadocs.org/docsets/CocoaLumberjack/)
* [x] I have searched for a similar pull request in the [project](https://github.com/CocoaLumberjack/CocoaLumberjack/pulls) and found (n)one (the one I found was for beta 6, not GM)

* [x] I have updated this branch with the latest master to avoid conflicts (via merge from master or rebase)
* [x] I have added the required tests to prove the fix/feature I am adding
* [x] I have updated the documentation (if necesarry)
* [x] I have run the tests and they pass
* [x] I have run the lint and it passes (`pod lib lint`)

This merge request fixes / reffers to the following issues: Adds Swift 3.0 Support

### Pull Request Description

I've updated all Swift Sources for Swift 3.0. I'd recommend merging this into a separate branch to maintain Swift 2.3 compatibility on master. Or create a new branch from the current master state and use it as support for Swift 2.3 while merging this into master.

### CODE REVIEW

**Comment by chrisdoc** [comment]
@ffried could you please change `osx_image` from `xcode7.3` to `xcode8` in https://github.com/ffried/CocoaLumberjack/blob/swift3.0/.travis.yml so that Travis can compile Swift 3 code

**Comment by ffried** [comment]
@chrisdoc Sure. Done.

**Comment by BarakRL** [comment]
@ffried can you please change AnyObject to Any in CocoaLumberjack.swift ?
`(... tag: AnyObject?, ...)`
to
`(... tag: Any?, ...)`

(there are about 7 changes in the file.)

**Comment by ffried** [comment]
@BarakRL Done as well.

**Comment by BarakRL** [comment]
lint fails because of a bug in cocoapods and xcode 8, see: https://github.com/CocoaPods/CocoaPods/issues/5702

Can someone merge this pull request please?

**Comment by hons82** [comment]
I've just made a PR on @ffried fork, this will set the language version to Swift 3.0 on a pod install/update, so there is no need of having a post installer.
Maybe there is something different planned by Cocoapods, however right now I just found this solution or you'll have to add a post install step

**Comment by danielgalasko** [comment]
Also feel like these PRs should in the very least add support for Swift 2.3 as well to make the transition easier for consumers

**Comment by ffried** [comment]
@danielgalasko That's why I suggested to create a separate branch for Swift3. Or one for Swift2.3. Most projects will probably sooner or later convert to Swift 3. I don't know if and how many changes are necessary for Swift 2.3 AFAIK the current master state already supports it. Which means a new Swift 2.3 branch could be created from the current master state and then this PR could be merged into master.

**Comment by hons82** [comment]
Actually this PR comes from a `swift3.0` branch.
In my opinion it would be perfectly suited in this case to have more than a single branch... if it is for a limited time

**Comment by ffried** [comment]
@hons82 Yea. Like I said it's also a possibility to have a swift3.0 branch for the time being.

Personally, I've got the impression that most projects update directly to Swift 3, which would make it easier for those projects to have Swift 3 support in master and 2.3 in its own branch.
However I might be wrong...  😃 

Having Swift 2.3 on a separate branch would also clearly mark it as "legacy support" since all new development would primarily happen on the master branch in Swift 3. That's more or less what Swift 2.3's purpose is according to Apple.

**Comment by danielgalasko** [comment]
Totally @ffried we moving to Swift 3 directly. Was merely saying that since its possible to support both I dont see why we shouldn't:) Anything I can help with to get this stuff merged? Not sure what we waiting on?

**Comment by devxoul** [comment]
You may need CocoaPods >= 1.1.0.rc.x and file named `.swift-version` to push Swift 3 pods properly.

```console
$ echo 3.0 > .swift-version
```

File: Classes/CocoaLumberjack.swift [hunk 1]
```diff
@@ -16,7 +16,7 @@
 import Foundation
 
 extension DDLogFlag {
-    public static func fromLogLevel(logLevel: DDLogLevel) -> DDLogFlag {
+    public static func fromLogLevel(_ logLevel: DDLogLevel) -> DDLogFlag {
```
Comment (devxoul) [inline Classes/CocoaLumberjack.swift]: How about renaming this in `from(_ logLevel: DDLogLevel)` to follow Swift 3 API naming convention?

File: Framework/iOSSwift/Formatter.swift [hunk 2]
```diff
@@ -10,30 +10,30 @@ import Foundation
...
+        threadUnsafeDateFormatter = DateFormatter()
+        threadUnsafeDateFormatter.formatterBehavior = .behavior10_4
         threadUnsafeDateFormatter.dateFormat = "HH:mm:ss.SSS"
         
         super.init()
     }
     
-    override func formatLogMessage(logMessage: DDLogMessage!) -> String {
-        let dateAndTime = threadUnsafeDateFormatter.stringFromDate(logMessage.timestamp)
+    override func formatLogMessage(_ logMessage: DDLogMessage!) -> String {
```
Comment (devxoul) [inline Framework/iOSSwift/Formatter.swift]: My proposal:

```diff
- override func formatLogMessage(_ logMessage: DDLogMessage!) -> String
+ override func format(_ logMessage: DDLogMessage!) -> String
```

File: Framework/watchOSSwiftTest Extension/Formatter.swift [hunk 3]
```diff
@@ -10,30 +10,30 @@ import Foundation
...
+        threadUnsafeDateFormatter = DateFormatter()
+        threadUnsafeDateFormatter.formatterBehavior = .behavior10_4
         threadUnsafeDateFormatter.dateFormat = "HH:mm:ss.SSS"
         
         super.init()
     }
     
-    override func formatLogMessage(logMessage: DDLogMessage!) -> String {
-        let dateAndTime = threadUnsafeDateFormatter.stringFromDate(logMessage.timestamp)
+    override func formatLogMessage(_ logMessage: DDLogMessage!) -> String {
```
Comment (devxoul) [inline Framework/watchOSSwiftTest Extension/Formatter.swift]: Same here

**Comment by ffried** [comment]
@devxoul Fair point! However, since the core of CocoaLumberjack is still written in Objective-C, those changes would have to be made by applying `NS_SWIFT_XXX` to the Objective-C headers (except for `from(_: DDLogLevel)`). For the current changes I just used Xcode's own migrator to convert to Swift 3.

I'll see if I can find some time this weekend to play around with the new attributes and add them to the Objective-C headers so that the Swift version of it follows the new naming conventions more closely.

**Comment by ffried** [comment]
Just had some spare time and went through the header files (hope I didn't miss anything important).

Apart from adding `NS_SWIFT_NAME(...)` to some methods and properties, I've also changed some methods to readonly properties (but only if I was sure that nothing will break 😃). Also, I've annotated all `NSArray`'s and `NSDictionary`'s I've come across with what they contain.
Let me know if I overlooked something.

What I didn't do is to add nullability specifiers. For that I simply don't have enough knowledge about the core framework to know what can and cannot ever be `nil`. If someone wants to open a PR on my repo with added nullability specifiers, I'd be happy to merge it.

Maybe I'll have some time in the near future to add those myself, but I'd rather open a new PR for this since it would delay this one too much.

**Comment by ffried** [comment]
Just increased the version to 2.4.0 so that only the tag needs to be created once merged.
I've also just installed cocoapods 1.1rc and this PR also passes validation:
pod lib lint result

File: CocoaLumberjack.podspec [hunk 4]
```diff
@@ -2,7 +2,7 @@
 Pod::Spec.new do |s|
 
   s.name     = 'CocoaLumberjack'
-  s.version  = '2.3.0'
+  s.version  = '2.4.0'
```
Comment (max-potapov) [inline CocoaLumberjack.podspec]: Swift 3.0 has a breaking API changes. I'll suggest to use `3.0.0` version instead.

Comment (ffried) [inline CocoaLumberjack.podspec]: Sure. I just saw that adding support for Swift 2.2 also lead to a minor version increase. I'll change this to 3.0.0 (kinda matches the Swift version 😄)

**Comment by ffried** [comment]
Commit 898bf1c4e35d88a7ca0efb85f3e0bfaff5e8cc6d installs CocoaPods version 1.1rc for Travis. It should be reverted as soon as v1.1 is released.

**Comment by ffried** [comment]
The travis build is still failing, but I don't know what's wrong. The same command runs fine on my local machine. But the `pod lib lint` command now runs fine. I've just upgraded the project settings and schemes of the Tests project. Maybe that resolves it.

Anyway. From my point of view this PR is now ready to be merged. Or did I miss anything?

File: Classes/CocoaLumberjack.swift [hunk 5]
```diff
@@ -26,77 +26,75 @@ extension DDLogFlag {
...
+public var defaultDebugLevel = DDLogLevel.verbose
 
 public func resetDefaultDebugLevel() {
-    defaultDebugLevel = DDLogLevel.Verbose
+    defaultDebugLevel = DDLogLevel.verbose
 }
 
-@available(*, deprecated, message="Use one of the DDLog*() functions if appropriate or call _DDLogMessage()")
-public func SwiftLogMacro(isAsynchronous: Bool, level: DDLogLevel, flag flg: DDLogFlag, context: Int = 0, file: StaticString = #file, function: StaticString = #function, line: UInt = #line, tag: AnyObject? = nil, @autoclosure string: () -> String, ddlog: DDLog = DDLog.sharedInstance()) {
+@available(*, deprecated, message: "Use one of the DDLog*() functions if appropriate or call _DDLogMessage()")
```
Comment (max-potapov) [inline Classes/CocoaLumberjack.swift]: Should we remove deprecated methods on major version upgrade?

Comment (kballard) [inline Classes/CocoaLumberjack.swift]: I think this has been deprecated for a while, so it seems reasonable to me to remove it, though it doesn't really hurt to leave it there.

Comment (ffried) [inline Classes/CocoaLumberjack.swift]: Actually, I've stumbled across this one as well. Since we're already three then, I've removed it. 👍

**Comment by max-potapov** [comment]
@bpoplauschi @dvor @rivera-ernesto 
Hi, guys! Could you please review this changes?

**Comment by max-potapov** [comment]
@ffried PR is confliction with master due to prior changes for Swift 2.3

**Comment by ffried** [comment]
@max-potapov Is there a way I can see what exactly the conflicts are? Github doesn't seem to show me.
I guess it's just the changes that pin the targets to a Swift version. These would have to be overwritten by this PR for Swift 3 to work.
Again, I'd recommend to create a `swift2.3` branch from the current `master` state and then merge this PR so that both versions of Swift are supported.

**Comment by max-potapov** [comment]
@ffried Sure!
1. fetch master from upstream 
2a. merge upstream/master to your swift 3 branch, resolve conflicts and then push with merge commit.
or
2b. rebase your swift 3 branch to upstream/master, resolve conflicts and then force push your branch.

**Comment by ffried** [comment]
@max-potapov Ah, ok. Thanks! Thought there'd be a way to check this online. :)

**Comment by max-potapov** [comment]
@ffried 2.3 can be branched anytime if needed.

**Comment by max-potapov** [comment]
@bpoplauschi May I ask you to merge this PR?

**Comment by devxoul** [comment]
:tada:

**Comment by bpoplauschi** [comment]
Wow guys, I'm really impressed by this PR. Everything looks right, especially bumping the version to 3.0.0 since there are breaking changes, tests pass, all demo projects are updated, Swift renaming, ... Very nice job. Many thanks to @ffried @max-potapov @chrisdoc @BarakRL @devxoul and all the others that contributed. I will test is a bit and release a 3.0.0 version via CocoaPods asap.

**Comment by bpoplauschi** [comment]
I just released 3.0.0 version via CocoaPods.

**Pull request merged**